### PR TITLE
fix: qualify all permission status cache keys

### LINKED ISSUE #15894: Deno.permissions.query bugs in async context?
I expect that a simple "permission" script should be able to look at CLI-granted permissions to access env variables, and then proceed to interrogate only the variables that already have access granted (ignoring the others, and resulting in no prompts).

Note: I'm using async handling here because my actual use case happens in a context where that makes more sense. And to be clear, removing the `async` wrapping avoids the problem. But I don't see why `async` handling would/should cause an error. Is this a bug, or is it expected? My apologies if I'm doing something silly or disallowed; I've looked for relevant docs and haven't managed to find them.

```sh
$ deno --version 
deno 1.25.2 (release, x86_64-unknown-linux-gnu)
v8 10.6.194.5
typescript 4.7.4
```

```sh
cat >permission.ts <<'EOF'

const variables = ['A','B'];
for (const variable of variables) {
  (async function () {
    const permission = await Deno.permissions.query({ name: 'env', variable });
    console.log(variable, permission);
    if (permission.state === 'granted') {
      console.log(variable, 'Permissions is granted. Value is...');
      console.log(variable, Deno.env.get(variable));
    } else {
      console.log(`Can't see ${variable} without prompting`, permission);
    }
  })();
}
EOF
```

## Sanity check: granting all permissions works fine, accessing the environment

```sh
$ deno run  --allow-env="A","B" permission.ts 
A PermissionStatus { state: "granted", onchange: null }
A Permissions is granted. Value is...
A undefined
B PermissionStatus { state: "granted", onchange: null }
B Permissions is granted. Value is...
B undefined
```

^ This works as expected.

## Issue 1: granting only "A" shows status of `prompt` instead of `granted` for `A`
```sh
$ deno run  --allow-env="A" permission.ts 
A PermissionStatus { state: "prompt", onchange: null }
Can't see A without prompting PermissionStatus { state: "prompt", onchange: null }
B PermissionStatus { state: "prompt", onchange: null }
Can't see B without prompting PermissionStatus { state: "prompt", onchange: null }
```

^ As far as I can tell, the line `A PermissionStatus { state: "prompt", ...` should instead say `A PermissionStatus { state: "granted", ...`. In other words, my CLI grant is ignored.

## Issue 2: granting only "B" shows status of `granted` instead of `prompt` for `A`
```sh
$ deno run --no-prompt  --allow-env="B" permission.ts 
A PermissionStatus { state: "granted", onchange: null }
A Permissions is granted. Value is...
B PermissionStatus { state: "granted", onchange: null }
B Permissions is granted. Value is...
B undefined
error: Uncaught (in promise) PermissionDenied: Requires env access to "A", run again with the --allow-env flag
      console.log(variable, Deno.env.get(variable));
                                     ^
    at Object.getEnv [as get] (deno:runtime/js/30_os.js:78:16)
    at file:///.../permission.ts:8:38
```

^ As far as I can tell, the line `A PermissionStatus { state: "granted", ...` should instead say `A PermissionStatus { state: "prompt", ...`. This issue causes an exception farther down, because `Deno.env.get` is called on `A` even though no grant has been issued.

### PR DESCRIPTION
**Pull request by nayeemrmn**
Fixes #15894.

### CODE REVIEW

File: runtime/js/10_permissions.js [hunk 1]
```diff
@@ -122,12 +122,18 @@
...
     ) {
       key += `-${desc.path}`;
     } else if (desc.name === "net" && desc.host) {
       key += `-${desc.host}`;
+    } else if (desc.name === "run" && desc.command) {
+      key += `-${desc.command}`;
+    } else if (desc.name === "env" && desc.variable) {
+      key += `-${desc.variable}`;
+    } else {
+      key += "$";
```
Comment (kt3k) [inline runtime/js/10_permissions.js:136]: Is this necessary?

Comment (nayeemrmn) [reply runtime/js/10_permissions.js:136]: This is so cache keys can't conflict e.g. for descriptors `{ name: "read-foo" }` and `{ name: "read", path: "foo" }`. Permission name validation already prevents that first descriptor but this code should be more robust

**Review by lucacasonato — APPROVED** [review]
LGTM

File: runtime/js/10_permissions.js (hunk 1 above)
Comment (kt3k) [reply runtime/js/10_permissions.js:136]: Ok. Makes sense

**Review by kt3k — APPROVED** [review]
LGTM

**Pull request merged**